### DISCUSSION TITLE
Intercepting commands and responses by supplying a custom protocol object in the Jedis Client

### DISCUSSION THREAD

**Discussion by gaspardpetit** [xetorthio/jedis#2635] [Ideas]
Hi, I would like to know if there would be interest to allow customization of the `Protocol` currently used by a `Connection`. For context, the three main use case we have are

 - enabling logs of individual redis commands on our client application
 - monitoring slow response from redis so they can be logged as warnings
 - making sure that all errors from redis are reported properly

While this kind of monitoring can be done relatively easily from the redis instance itself, it is often preferable to do it on the client because it can provide a lot of details about the context helping understand what cause the command.

I could not find a way to enable this kind of logging, and the easiest change I could come up with replaces the usage of the `Protocol` static methods to a `JedisProtocol` instance that calls the `Protocol` static methods, but can also be substituted when creating the `Client`.  This allows the client application to provide a subclass of the `JedisProtocol` with special handling around execution, response and error handling.

The `JedisProtocol` would expose the three methods needed to achieve these use cases:

```.java
public class JedisProtocol {
  public static final JedisProtocol DEFAULT = new JedisProtocol();

  public void sendCommand(final RedisOutputStream os, final ProtocolCommand command,
      final byte[]... args) {
    Protocol.sendCommand(os, command, args);
  }

  public String readErrorLineIfPossible(RedisInputStream is) {
    return Protocol.readErrorLineIfPossible((is));
  }

  public Object read(RedisInputStream is) {
    return Protocol.read(is);
  }
}
```

This idea is presented as a pull request (https://github.com/redis/jedis/pull/2633) let me know if there is interest, I am happy to adjust change if needed.

**Comment by yangbodong22011** [comment]
@gaspardpetit nice idea.
